### PR TITLE
[issue - 28, 67, 68, 72,73] improvements in bitbucket subscriptions 

### LINKED ISSUE #28: Security: Regular User is allowed to view and manage list of subscriptions belonging to a different user
From GH Comment: https://github.com/mattermost/mattermost-plugin-bitbucket/pull/5#pullrequestreview-626014445

##### 6) Regular User is allowed to view and manage list of subscriptions belonging to a different user.

Severity: Low

Steps:

* Login as User1 on Mattermost.
* Connect to Bitbucket as User1.
* Subscribe to few private repositories using the command /bitbucket subscribe user1/repo1
* On another browser, login as User2 with low priviliges and visit the same channel.
* Check subscriptions /bitbucket subscribe list and notice that it still displays repo1 which is a private repo of user1.
* Unsubscribe using the command /bitbucket unsubscribe user1/repo1 and notice that user is allowed to change the subscription belonging to a different user.

Expected: Subscriptions should be user based. Only the owner of the subscription should be allowed to view or unsubscrib

### LINKED ISSUE #67: Without /bitbucket connect, /bitbuket help Command Should Work. same as Github. (improvement)
Currently when the user use /bitbucket help then his getting `You must connect your account to Bitbucket first. Either click on the Bitbucket logo in the bottom left of the screen or enter /bitbucket connect.`  but same as github if plugin is Working and Config is Right then user Should able to use help Command without using /bitbucket Connect.

### LINKED ISSUE #68: Without Subscribe repo user can Unsubscribe any repo and user also getting the success message.
issue :- Without subscribe any repo if user type /bitbucket subscription delete `you can type anything here` and press enter then user will get success message and unsubscribed repo is deleted.
image

the same issue is there in the Github, issue link :- https://github.com/mattermost/mattermost-plugin-github/issues/546

### LINKED ISSUE #72: Subscription creation shows both an ephemeral and a normal post to the channel
When a user creates a Bitbucket subscription, a post is shown with subscription info for all channel members to see. 

The issue is that the ephemeral post is also shown to the user who created the subscription. This ephemeral post is redundant and not needed. 

Steps:
- Setup Bitbucket and connect a user
- Add a subscription to a channel 
**Observed**: 
image

### LINKED ISSUE #73: User see's an incorrect create an invalid subscription if they attempt to subscribe but, provide only an owner
If a user attempts to add a subscription with only an owner but no repo, they will see a success message and a non-functional subscription is created. Bitbucket does not appear to support any subscriptions that target only an owner. Subscriptions must be in the `owner/repo` format.

Steps:
- Setup Bitbucket and connect a user
- Add a subscription to a channel that is only for an owner that does not exist: `/bitbucket subscriptions add apple-does-not-exist`

**Observed**:
image

**Expected**: User should be forced to also supply a repo. The validation that exists will then check if the repo is valid

### PR DESCRIPTION
**Pull request by sibasankarnayak**
made some improvements in logic for subscription

Ticket Here
Fixes #67 
Fixes #68 
Fixes #72 
Fixes #73

### CODE REVIEW

**Comment by sibasankarnayak** [comment]
@dipak-demansol Can you start the QA in mean time to crosscheck if i had covered all the logical cases.

**Review by mickmister — COMMENTED** [review]
@sibasankarnayak Can you please remove the changes pertaining to attaching subscription to its creator? It was determined in https://github.com/mattermost/mattermost-plugin-bitbucket/issues/28 that this new behavior is undesired.

As long as "regular" users cannot see or edit subscriptions in channels they are not a member of, the current behavior is fine.

**Comment by dipak-demansol** [comment]
@sibasankarnayak  i'll start working on after you Complete the changes as per mickmister Comment.

**Comment by sibasankarnayak** [comment]
> dit subscriptions in channels they are not a member of, the current behavior is fine.



> @sibasankarnayak Can you please remove the changes pertaining to attaching subscription to its creator? It was determined in #28 that this new behavior is undesired.
> 
> As long as "regular" users cannot see or edit subscriptions in channels they are not a member of, the current behavior is fine.

@mickmister Sure, have seperated the code for issue #28

@dipak-demansol you can start testing

**Review by dipak-demansol — CHANGES_REQUESTED** [review]
@sibasankarnayak 
:- https://github.com/mattermost/mattermost-plugin-bitbucket/issues/67 :- LGTM 
:- https://github.com/mattermost/mattermost-plugin-bitbucket/issues/72 :- LGTM
:- https://github.com/mattermost/mattermost-plugin-bitbucket/issues/73 :- LGTM
:- https://github.com/mattermost/mattermost-plugin-bitbucket/issues/68 :- have minor issue, issue link https://drive.google.com/file/d/1A68jP9Ig-6xjP8ejMfBxY2SBNCKDFQFx/view?usp=sharing

**Comment by sibasankarnayak** [comment]
> @sibasankarnayak :- #67 :- LGTM :- #72 :- LGTM :- #73 :- LGTM :- #68 :- have minor issue, issue link https://drive.google.com/file/d/1A68jP9Ig-6xjP8ejMfBxY2SBNCKDFQFx/view?usp=sharing

@dipak-demansol can you retest once more , fixed the logic

**Review by dipak-demansol — APPROVED** [review]
LGTM

File: server/command.go [hunk 1]
```diff
@@ -227,28 +228,7 @@ func (p *Plugin) handleSubscribe(_ *plugin.Context, args *model.CommandArgs, par
...
-				Message:   msg,
-			}
-
-			if _, appErr := p.API.CreatePost(post); appErr != nil {
-				p.API.LogWarn("error while creating post", "post", post, "error", appErr.Error())
-				return fmt.Sprintf("%s Though there was an error creating the public post: %s", msg, appErr.Error())
-			}
-
-			return msg
+			return requiredErrorMessage
```
Comment (mickmister) [inline server/command.go:231]: It looks like we can't subscribe to orgs anymore. Is this intentional?

File: server/command.go [hunk 2]
```diff
@@ -273,7 +253,7 @@ func (p *Plugin) handleSubscribe(_ *plugin.Context, args *model.CommandArgs, par
 			return fmt.Sprintf("%s Though there was an error creating the public post: %s", msg, appErr.Error())
 		}
 
-		return msg
+		return ""
```
Comment (mickmister) [inline server/command.go:256]: Why not return `msg`?

File: server/subscriptions.go [hunk 3]
```diff
@@ -195,21 +195,25 @@ func (p *Plugin) GetSubscribedChannelsForRepository(pl webhookpayload.Payload) [
...
-func (p *Plugin) Unsubscribe(channelID string, repo string) error {
+func (p *Plugin) Unsubscribe(channelID, repo string) (string, error) {
+	if len(strings.Split(repo, "/")) != 2 {
+		return requiredErrorMessage, nil
+	}
+
 	owner, repo := parseOwnerAndRepo(repo, p.getBaseURL())
 	if owner == "" && repo == "" {
-		return errors.New("invalid repository")
+		return "", errors.New("invalid repository")
```
Comment (mickmister) [inline server/subscriptions.go]: Would this work here as well?

```suggestion
		return requiredErrorMessage, nil
```

File: server/subscriptions.go [hunk 4]
```diff
@@ -195,21 +195,25 @@ func (p *Plugin) GetSubscribedChannelsForRepository(pl webhookpayload.Payload) [
...
 	subs, err := p.GetSubscriptions()
 	if err != nil {
-		return errors.Wrap(err, "could not get subscriptions")
+		return "", errors.Wrap(err, "could not get subscriptions")
 	}
 
 	repoSubs := subs.Repositories[repoWithOwner]
 	if repoSubs == nil {
-		return nil
+		return fmt.Sprintf("Unable to unsubscribed from %s as it is not subscribed.", repo), nil
```
Comment (mickmister) [inline server/subscriptions.go]:
```suggestion
		return fmt.Sprintf("Unable to unsubscribe from %s as it is not currently part of a subscription in this channel.", repo), nil
```

File: server/subscriptions.go [hunk 5]
```diff
@@ -224,9 +228,10 @@ func (p *Plugin) Unsubscribe(channelID string, repo string) error {
...
 		subs.Repositories[repoWithOwner] = repoSubs
 		if err := p.StoreSubscriptions(subs); err != nil {
-			return errors.Wrap(err, "could not store subscriptions")
+			return "", errors.Wrap(err, "could not store subscriptions")
 		}
+		return fmt.Sprintf("Successfully unsubscribed from %s.", repo), nil
 	}
 
-	return nil
+	return fmt.Sprintf("Unable to unsubscribed from %s as it is not subscribed.", repo), nil
```
Comment (mickmister) [inline server/subscriptions.go]:
```suggestion
	return fmt.Sprintf("Unable to unsubscribe from %s as it is not currently part of a subscription in this channel.", repo), nil
```

**Review by mickmister — CHANGES_REQUESTED** [review]

File: server/command.go (hunk 1 above)
Comment (sibasankarnayak) [reply server/command.go:231]: @mickmister yes can see the comment https://github.com/mattermost/mattermost-plugin-bitbucket/issues/73#issue-1175540416

File: server/command.go (hunk 2 above)
Comment (sibasankarnayak) [reply server/command.go:256]: @mickmister can get the context here behind this logic https://github.com/mattermost/mattermost-plugin-github/pull/517#issuecomment-989178101 , later on this decided to show only a public post on successful subscription

File: server/subscriptions.go [hunk 6]
```diff
@@ -195,21 +195,25 @@ func (p *Plugin) GetSubscribedChannelsForRepository(pl webhookpayload.Payload) [
...
 	subs, err := p.GetSubscriptions()
 	if err != nil {
-		return errors.Wrap(err, "could not get subscriptions")
+		return "", errors.Wrap(err, "could not get subscriptions")
 	}
 
 	repoSubs := subs.Repositories[repoWithOwner]
 	if repoSubs == nil {
-		return nil
+		return fmt.Sprintf("Unable to unsubscribe from %s as it is not currently part of a subscription in this channel.", repo), nil
```
Comment (mickmister) [inline server/subscriptions.go]: Can we put this string at the top of the file as a constant?

File: server/command.go [hunk 7]
```diff
@@ -435,6 +408,14 @@ func (p *Plugin) ExecuteCommand(c *plugin.Context, args *model.CommandArgs) (*mo
 		return &model.CommandResponse{}, nil
 	}
 
+	if action == "help" {
+		if f, ok := p.CommandHandlers[action]; ok {
+			message := f(c, args, parameters, nil)
+			p.postCommandResponse(args, message)
+			return &model.CommandResponse{}, nil
+		}
+	}
```
Comment (mickmister) [inline server/command.go]: We can just call the help function directly here, instead of accessing `CommandHandlers`

**Review by mickmister — APPROVED** [review]

**Review by jfrerich — APPROVED** [review]

**Comment by hanzei** [comment]
@mickmister @jfrerich Can we merge this PR? (After resolving the conflicts)

**Comment by jfrerich** [comment]
@hanzei, I don't see why we wouldn't.

**Comment by mickmister** [comment]
> Can we merge this PR? (After resolving the conflicts)

@hanzei This PR was submitted by Demansol. I'm not sure if we're able to update the PR with the conflict resolutions

**Comment by hanzei** [comment]
Re-submitted as https://github.com/mattermost/mattermost-plugin-bitbucket/pull/97

**Pull request closed without merging**